### PR TITLE
Add instructions for arbitrary sized byte arrays

### PR DESCRIPTION
**Pull request by hanssv**

### CODE REVIEW

**Review by ghallak — DISMISSED** [review]

**Review by ghallak — APPROVED** [review]

File: src/aeb_fate_generate_ops.erl [hunk 1]
```diff
@@ -247,6 +247,12 @@ ops_defs() ->
     , { 'DBG_UNDEF',           16#b3,  false,    true,   true,    ?GAS(0),       [a, a],     dbg_undef,               {string, any},    none, "Debug Op: Undefine a variable. Args = {var_name, register}" }
     , { 'DBG_CONTRACT',        16#b4,  false,    true,   true,    ?GAS(0),          [a],  dbg_contract,                    {string},    none, "Debug Op: Name the current contract. Args: {contract_name}"}
 
+    , { 'BYTES_SPLIT_ANY',     16#b5,  false,    true,   true,   ?GAS(10), [a, a, a],     bytes_split_any,   {bytes, integer}, variant, "Arg0 := bytes_split_any(Arg1, Arg2), where a positive Arg2 is the length of the first chunk, and a negative Arg2 is the length of the second chunk. Returns None if byte array is not long enough."}
```
Comment (radrow) [inline src/aeb_fate_generate_ops.erl:250]: Can you squeeze it below debugging instructions?

Comment (radrow) [inline src/aeb_fate_generate_ops.erl:250]: Why arg2 has to be negative?

File: src/aeb_fate_generate_ops.erl [hunk 2]
```diff
@@ -247,6 +247,12 @@ ops_defs() ->
...
+    , { 'BYTES_SIZE',          16#b6,  false,    true,   true,   ?GAS(10),    [a, a],          bytes_size,            {bytes}, integer, "Arg0 := bytes_size(Arg1), returns the number of bytes in the byte array."}
+    , { 'BYTES_TO_FIXED_SIZE', 16#b7,  false,    true,   true,   ?GAS(10), [a, a, a], bytes_to_fixed_size,   {bytes, integer}, variant, "Arg0 := bytes_to_fixed_size(Arg1, Arg2), returns Some(Arg1) if byte_size(Arg1) == Arg2, None otherwise."}
```
Comment (radrow) [inline src/aeb_fate_generate_ops.erl:252]: How is it better than returning a bool? You still need to provide Arg1, so there is no point in getting it back. Also the name of the instruction is not descriptive.

File: src/aeb_fate_encoding.erl [hunk 3]
```diff
@@ -270,8 +272,12 @@ deserialize_type(<<?TYPE_TUPLE, N, Rest/binary>>) ->
     {{tuple, Ts}, Rest2};
 deserialize_type(<<?TYPE_BYTES, Rest/binary>>) ->
     {N, Rest2} = deserialize_one(Rest),
-    true       = is_integer(N) andalso N >= 0,
-    {{bytes, N}, Rest2};
+    true       = is_integer(N),
+    if N == -1 ->
```
Comment (radrow) [inline src/aeb_fate_encoding.erl:276]: Do any docs actually refer to this specifically chosen -1?

File: src/aeb_fate_generate_ops.erl (hunk 1 above)
Comment (hanssv) [reply src/aeb_fate_generate_ops.erl:250]: It doesn't have to be negative, it _can_ be negative. The alternative would be a three argument function, I thought this was more compact.

File: src/aeb_fate_generate_ops.erl (hunk 2 above)
Comment (hanssv) [reply src/aeb_fate_generate_ops.erl:252]: You have something of type `bytes()` and want something of type `bytes(n)` then a boolean wont help you? The `Arg1` has type `bytes(Arg2)` - perhaps the explanation can be improved...

File: src/aeb_fate_generate_ops.erl (hunk 1 above)
Comment (hanssv) [reply src/aeb_fate_generate_ops.erl:250]: Yes, sure!

File: src/aeb_fate_encoding.erl (hunk 3 above)
Comment (hanssv) [reply src/aeb_fate_encoding.erl:276]: Not yet - but it should go into `protocol` at some point.

File: src/aeb_fate_generate_ops.erl (hunk 2 above)
Comment (radrow) [reply src/aeb_fate_generate_ops.erl]: Ah, so it's casting! Makes sense

File: src/aeb_fate_generate_ops.erl (hunk 1 above)
Comment (radrow) [reply src/aeb_fate_generate_ops.erl]: I misread, makes sense now

**Review by radrow — APPROVED** [review]

**Review by ghallak — APPROVED** [review]

**Pull request merged**